fs/metrics/manager/monitors/image_monitor.go

```diff
@@ -0,0 +1,100 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package monitors
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	cm "github.com/awslabs/soci-snapshotter/fs/metrics/common"
+
+	"github.com/awslabs/soci-snapshotter/fs/layer/operations"
+	"github.com/containerd/containerd/log"
+	"github.com/opencontainers/go-digest"
+)
+
+// NewImageMonitor returns a new ImageMonitor. An ImageMonitor
+// encapsulates the emission of FUSE telemetry data at an image
+// level.
+func NewImageMonitor(waitP time.Duration) *ImageMonitor {
+	return &ImageMonitor{
+		OpCounter:  make(map[digest.Digest]*fuseOperationCounter),
+		waitPeriod: waitP,
+	}
+}
+
+type ImageMonitor struct {
+	// opCounter maintains a fuseOperationCounter per image.
+	OpCounter  map[digest.Digest]*fuseOperationCounter
+	waitPeriod time.Duration
+}
+
+func (im *ImageMonitor) AddImageFuseOperationCount(operation string, image digest.Digest, count int32) {
+	cm.AddImageOperationCount(operation, image, count)
+}
+
+// AddOpCounterForImage constructs a FuseOperationCounter for an image
+// with digest imgDigest.
+func (im ImageMonitor) AddOpCounterForImage(imgDigest digest.Digest) {
+	f := &fuseOperationCounter{
+		opCounts: make(map[string]*int32),
+	}
+	for _, m := range operations.FuseOpsList {
+		f.opCounts[m] = new(int32)
+	}
+	im.OpCounter[imgDigest] = f
+}
+
+// RunOpCounter waits for a waitPeriod to pass before emitting a log and metric for each
+// operation in FuseOpsList. Should be started in different goroutine so that it
+// doesn't block the current goroutine.
+func (im *ImageMonitor) RunOpCounter(ctx context.Context, imgDigest digest.Digest) {
```

Comment:
Since we maintain a map of image digests to counters I thought it made sense to logically separate `Add` and `Run`, although like you mention, I don't see a case where we `Add` and don't `Run` immediately.